executor/index_advise.go

```diff
@@ -0,0 +1,146 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor
+
+import (
+	"context"
+	"math"
+	"strings"
+	"unsafe"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/parser"
+	"github.com/pingcap/parser/ast"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/util/chunk"
+)
+
+// IndexAdviseExec represents a index advise executor.
+type IndexAdviseExec struct {
+	baseExecutor
+
+	IsLocal         bool
+	indexAdviseInfo *IndexAdviseInfo
+}
+
+// Next implements the Executor Next interface.
+func (e *IndexAdviseExec) Next(ctx context.Context, req *chunk.Chunk) error {
+	req.GrowAndReset(e.maxChunkSize)
+	if !e.IsLocal {
+		return errors.New("Index Advise: don't support load data without local field")
+	}
+	if e.indexAdviseInfo.Path == "" {
+		return errors.New("Index Advise: infile path is empty")
+	}
+
+	if val := e.ctx.Value(IndexAdviseKey); val != nil {
+		e.ctx.SetValue(IndexAdviseKey, nil)
+		return errors.New("Index Advise: there is already an IndexAdviseExec running in the session")
+	}
+	e.ctx.SetValue(IndexAdviseKey, e.indexAdviseInfo)
+	return nil
+}
+
+// Close implements the Executor Close interface.
+func (e *IndexAdviseExec) Close() error {
+	return nil
+}
+
+// Open implements the Executor Open interface.
+func (e *IndexAdviseExec) Open(ctx context.Context) error {
+	return nil
+}
+
+// IndexAdviseInfo saves the information of index advise operation.
+type IndexAdviseInfo struct {
+	Path        string
+	MaxMinutes  uint64
+	MaxIndexNum *ast.MaxIndexNumClause
+	LinesInfo   *ast.LinesClause
+	Ctx         sessionctx.Context
+	StmtNodes   [][]ast.StmtNode
+	Result      *IndexAdvice
+}
+
+func (e *IndexAdviseInfo) getStmtNodes(data []byte) error {
+	str := *(*string)(unsafe.Pointer(&data))
```

Comment:
Can we use `string(data)` to convert it?

Reply:
Of course, but using `string (data)` will incur some overhead. After using the string, we will store the SQL statements as `[][]ast.StmtNode`, then we will never use this string again, so this method may be more suitable. 
And if this could cause some problems I didn't consider, I will modify it.

Reply:
If it is not the bottleneck, we should put code simplicity first.